### PR TITLE
feat: dialogs component registration

### PR DESCRIPTION
**Pull request by chon219**
## Description

Added dialogs component registration to provide memory scopes and path resolvers.

## Specific Changes

  - Added interface `ComponentMemoryScopes`
  - Added interface `ComponentPathResolvers`
  - Moved default memory scopes and path resolvers to `DialogsComponentRegistration`

## Testing

Updated related tests.

### CODE REVIEW

File: libraries/botbuilder-dialogs/src/memory/componentMemoryScopes.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * Check if a [ComponentRegistration](xref:botbuilder-core.ComponentRegistration) is
+ * [ComponentMemoryScopes](xref:botbuilder-dialogs.ComponentMemoryScopes) or not.
+ *
+ * @param {ComponentRegistration} component The component registration.
+ * @returns {boolean} Type check result.
+ */
+export function isComponentMemoryScopes(component: any): component is ComponentMemoryScopes {
+    return tests.isObject(component) && tests.isFunc(component['getMemoryScopes']);
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs/src/memory/componentMemoryScopes.ts:27]: If you want a fun(?) problem - try to write a `tests.unsafe.isObjectAs<T>(val: any): val is T` function that mirror the `assert.unsafe.castObjectAs<T>` function. It's a bit of a hack, but as long as it's combined with other tests/assertions I've found it to be helpful. Then you could write:

```typescript
export const isComponentMemoryScopes: Test<ComponentMemoryScopes> = (val) => {
  return tests.unsafe.isObjectAs<ComponentMemoryScopes>(component) && tests.isFunc(component.getMemoryScopes);
};
```

Not critical, just might be nice to have around.

File: libraries/botbuilder-dialogs/src/memory/index.ts [hunk 2]
```diff
@@ -5,10 +5,12 @@
  * Copyright (c) Microsoft Corporation. All rights reserved.
  * Licensed under the MIT License.
  */
-export * from './pathResolvers';
-export * from './scopes';
+export * from './componentMemoryScopes';
+export * from './componentPathResolvers';
 export * from './dialogPath';
 export * from './dialogStateManager';
+export * from './pathResolvers';
```
Comment (stevengum) [inline libraries/botbuilder-dialogs/src/memory/index.ts:12]: Thank you for ordering the exports 😄

File: libraries/botbuilder-dialogs/src/memory/componentMemoryScopes.ts (hunk 1 above)
Comment (chon219) [reply libraries/botbuilder-dialogs/src/memory/componentMemoryScopes.ts]: Thanks, I've tried the following implementation and it works, is it okay to include this in `botbuilder-stdlib`?
```
function isObjectAs<T>(val: unknown): val is T {
    if (isObject(val)) {
        castObjectAs<T>(val, []);
        return true;
    }
    return false;
}
```

File: libraries/botbuilder-stdlib/src/types.ts [hunk 3]
```diff
@@ -526,6 +526,25 @@ function makeTest<T>(assertion: Assertion<T>): Test<T> {
     };
 }
 
+/**
+ * **UNSAFE**
+ * Test if `val` is of type `object`.
+ * This assertion does not actually verify that `val` is of type `T`. It is useful as the first
```
Comment (joshgummersall) [inline libraries/botbuilder-stdlib/src/types.ts:532]: `assertion` should be `test`

**Review by joshgummersall — COMMENTED** [review]
Two quick suggestions, otherwise look good!

File: libraries/botbuilder-stdlib/src/types.ts [hunk 4]
```diff
@@ -526,6 +526,25 @@ function makeTest<T>(assertion: Assertion<T>): Test<T> {
...
+ * line in a nested assertion so that remaining assertion calls can leverage helpful intellisense.
+ * This method is only exported under the `unsafe` keyword as a constant reminder of this fact.
+ *
+ * @template T the type to cast `val` to, should extend `Dictionary<unknown>`, i.e. be itself an object
+ * @param {any} val value to test
+ * @returns {boolean} true if `val` is of type `object`
+ */
+function isObjectAs<T>(val: unknown): val is T {
+    if (isObject(val)) {
+        castObjectAs<T>(val, []);
```
[the remark this answers is not in the document]
Comment (chon219) [reply libraries/botbuilder-stdlib/src/types.ts:542]: So I can drop the `if` and just return the result of `isObject(val)`.

**Review by joshgummersall — APPROVED** [review]

**Pull request merged**